driver.go

```diff
@@ -47,26 +48,28 @@ import (
 // Driver is exported to allow it to be used directly.
 type Driver struct{}
 
-// Open a Connection to the server.
-// See https://github.com/apache/calcite-avatica-go#dsn for more information
-// on how the DSN is formatted.
-func (a *Driver) Open(dsn string) (driver.Conn, error) {
+// Connector implements the driver.Connector interface
+type Connector struct {
+	Infos  map[string]string
+	Client *http.Client
+
+	dsn string
+}
+
+// NewConnector creates a new connector
+func NewConnector(dsn string) driver.Connector {
+	return &Connector{nil, nil, dsn}
+}
 
-	config, err := ParseDSN(dsn)
+func (c *Connector) Connect(context.Context) (driver.Conn, error) {
+
+	config, err := ParseDSN(c.dsn)
 
 	if err != nil {
 		return nil, fmt.Errorf("Unable to open connection: %s", err)
 	}
 
-	httpClient, err := NewHTTPClient(config.endpoint, httpClientAuthConfig{
-		authenticationType:  config.authentication,
-		username:            config.avaticaUser,
-		password:            config.avaticaPassword,
-		principal:           config.principal,
-		keytab:              config.keytab,
-		krb5Conf:            config.krb5Conf,
-		krb5CredentialCache: config.krb5CredentialCache,
-	})
+	httpClient, err := NewHTTPClient(config.endpoint, c.Client)
```

Comment:
The logic from [the http client to set up the authentication ](https://github.com/apache/calcite-avatica-go/blob/master/http_client.go#L98) that was removed should be moved here.

It should use the `With...` methods to wrap the `http.Client` with authentication (if required) by looking at the `config` variable. We should only do this if `c.Client` is nil, so that it does not interfere with a user supplied client.

Reply:
done